Formats Class

eode opened the pull request:
~Title pretty much says it.  Someone else handled the case where metadata is present -- this is just infers the `target` when you're accessing a blob that doesn't have one.~

Sadly, at first blush this now looks like I've just replaced three functions and an Enum with three base classes, a couple of subclasses, and quite a few instances.  This is because I have.  But, there are justifications for all that, and benefits to it.

One of the original goals for this was to allow users to add  their own serializers/deserializers.  But there are a lot of things we'll need to *really* make that work.

User deserializers for filenames can be pretty simple, but the reality is that if you don't make a formal interface for a particular area of code, and give that formal interface structural authority over what's added / etc, then things get kinda lax around the edges.  Edge cases get implemented multiple times, or something that's generally needful gets implemented only in one area.  

So, this:
* Provides a generic class that can be instantiated easily to add a new serializer/deserializer pair
* handles extensions automatically for deserialization, if provided at format creation time
* handled class types automatically for serialization, if provided at format creation time
* can be subclassed for special cases, like where lazy loading is preferred
  * does that with numpy/pandas/arrow
* provides filename extensions that *should* be used for some object that is being serialized, but is getting a new name
* updates metadata (if given at serialization time) for round-trip serialization
* can be ripped out and unceremoniously plopped into Quilt4, when that happens.
  * including whatever developed/contributed `Format` subclasses or additional `Format` configured instances that have been added in the mean time.
* can be expanded later to perform chunk-iterative serialization for formats that support it, if that ever becomes a request.

Notes:
This replaces `serialize_obj()`, `deserialize_obj()`, `_get_target_for_object()` and the `TargetTypes` enum.  There is now no *static* definition of target types, but also, serialization/deserialization is isolated from other code, and in general it shouldn't be necessary to access the target type anywhere but from the `formats` module.

ToDo:
* [x] Complain about a specific format if specified but not registered.
* [x] Match ext on api.put() if metadata not given.
* [x] Handle pandas serialization
* [x] Fix parquet serialization compatibility with select()

eode commented:
Tests need fixing..

eode commented:
I'm going to make this a bit more user-facing.

eode commented:
Could still use some test expansion, but 's decent.

eode commented:
This could use a look -- @dimaryaz ?

eode commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,415 @@
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+
+        # don't include these extensions unlss numpy is present.
+        if package_exists('numpy'):
```
I have no problem with that.

eode replied:
Oh -- but, it wouldn't fail, in that case -- or shouldn't. `package_exists` doesn't detect whether it's imported, it detects whether it's installed, but without spending the cost of importing it.

eode replied:
Smart idea checking `sys.modules`, though.

eode commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,415 @@
+#! /usr/bin/python
```
IDE Template.  I'll strike that.

eode commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,415 @@
+
+This module handles binary formats, and conversion to/from objects.
+
+# Formats Class (plural)
```
Good call.

dimaryaz commented on api/python/t4/api.py:
```diff
@@ -69,7 +70,7 @@ def get(src):
     """
     data, meta = get_bytes(fix_url(src))
-    ext = pathlib.Path(src).suffix
+    ext = pathlib.PurePosixPath(src).suffix
```
Here, you actually want to use the result of `fix_url(src)`, which will always use `/`. Also, `urlparse(...).path`. It can have a query string like `versionId=...`.

eode replied:
Ah.  I thought we only accepted URLs as input, and didn't realize we were using query strings now.  S3 doesn't, from what I understand, and I didn't figure the filesystem would.  But if we're parsing them for our own use, that makes sense.  I probably should have accounted for the possibility anyways.

eode replied:
fixed.

eode commented on api/python/t4/api.py:
```diff
@@ -52,6 +52,7 @@ def put(obj, dest, meta=None):
         meta (dict): Optional. metadata dict to store with ``obj`` at ``dest``
     """
     all_meta = {'user_meta': meta}
+    ext = pathlib.PurePosixPath(dest).suffix
```
fixed.

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,418 @@
+        """Match a format (or formats) by extension."""
+        ext = ext.lower().strip('. ')
+        if single:
+            matching_formats = cls.formats_by_ext[ext]
```
Move it out of the `if`, since it's used in both cases

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,418 @@
+
+    @classmethod
+    def _get_format_name_from_meta(cls, meta):
+        name = None
```
Shortcut:
```
name = meta.get('format', {}).get('name')
```

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,418 @@
+            if getattr(self, '_deserializer', None) is None:
+                raise NotImplementedError()
+            format_meta = meta.get('format', {})
+            deserialization_kwargs = format_meta.get('deserialization', {})
```
I'm not sure this is a good idea... This could potentially allow things like `np.load(..., allow_pickle=True)` (ok, looks like you're not using kwargs for numpy and parquet, but still.) But also, it's very Python-specific.

eode replied:
The former -- fine, but the too-python-specific part I can get on board with.  I'll change it.

eode replied:
In consideration, even the former is rather an issue -- we shouldn't make it easy for mistakes to be made.

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,418 @@
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+
+        if 'parquet' not in handled_extensions:
+            handled_extensions.append('parquet')
```
Probably not good to modify caller's data... but really, I'd just hardcode the extensions for parquet and numpy for now.

dimaryaz commented on api/python/tests/test_formats.py:
```diff
@@ -0,0 +1,84 @@
+#! /usr/bin/python
```
Remove this one

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,418 @@
+        except AssertionError:
+            # Try again to convert the table after removing
+            # the possibly buggy Pandas-specific metadata.
+            # XXX: Can we detect this during serialization and store it separately?
```
Nope... it might not have been created by us. In any case, Pyarrow raising an assertion error on user input is a pyarrow bug - but they're not in a rush to fix it.

eode replied:
Thanks for the reply.

dimaryaz commented on api/python/t4/api.py:
```diff
@@ -51,13 +51,12 @@ def put(obj, dest, meta=None):
-    )
+    all_meta = {'user_meta': meta}
+    clean_dest = fix_url(dest)
+    ext = pathlib.PurePosixPath(urlparse(clean_dest).path).suffix
```
Actually, sorry, one more ridiculous URL-related bug: need `unquote(urlparse(clean_dest).path)`.

(E.g., `s3://bucket/foo.%74x%74` is the same URL as `s3://bucket/foo.txt`)

eode replied:
Ugh.  yeah, URLs should technically be percent-decoded and then utf-8 decoded -- but then, shouldn't we only accept bytes for URLs, since an encoded URL is a byte-format type?

eode replied:
..but no, we're not taking bytes here, we're taking a unicode-encoded object -- and we should use the best format for internal manipulation, and expect that format from our users -- particularly since we're going to be used by hand in (for example) Jupyter notebooks, ipython, bpython, etc.

stevededalus replied:
I'm not following this thread. Isn't `unquote` not changing the encoding unless you pass it an optional argument? It's just remapping url escaped characters.  It looks like the the bucket api class we keep s3 urls unescaped internally and we should be consistent about that.

But looking at the bucket api, it looks like these are never escaped when accessing a bucket? Is the s3manager expecting unescaped strings @dimaryaz ?  If so, why ever escape/unescape?

eode replied:
Unquote assumes that it's encoded as UTF-8 (which it's supposed to).  When quoting by spec, first the text is encoded as UTF-8, then encoded using percent encoding.  When unquoting, first the bytes are decoded using percent encoding, and then decoded via UTF-8.

The reason this is an issue is that we can't unquote more than once, ultimately.  We need to choose what our "external" or public-facing methods are, and what those accept.  I don't think this is a public-facing method.  If someone uses it, they should know what they're sending.

The only other consistent option is to do it right before needing to use the URL externally -- and that's problematic in our case because we do a lot of internal manipulation of the URLs.

I'd say that right now, we leave this as-is, and later convert everything over to an internal or external URI lib.  In the mean time, we *will* (and already do, in many areas) have inconsistencies with the URL spec.  `file://` when calling `Package.set()` is one example.

dimaryaz replied:
You can try a path with a space and see what happens, e.g.:
```
urllib.parse.urlparse(fix_url('a b.txt')).path
```
You'll end up with `%20` - so need to `unquote` it.

dimaryaz commented on api/python/t4/api.py:
```diff
@@ -71,17 +70,11 @@ def get(src):
-        raise QuiltException("No serialization metadata")
+    clean_src = fix_url(src)
+    data, meta = get_bytes(clean_src)
+    ext = pathlib.PurePosixPath(urlparse(clean_src).path).suffix
```
Same here

eode replied:
see comment above.

eode commented:
PR points addressed, re-merged with master (including integrating into `Bucket`).  I'm ready to merge this, pending approval of this one point:

The 'meta' dict is modified on serialization.  It could very reasonably be argued that info should instead be returned to the caller, and the calller should update that.  Instead, I took the stance that the `formats` module "owns" the `meta['format']` key.  This leads to shorter code when using it, but is more implicit.  The benefit is that it means less boilerplate on the caller's side.

..in any case, I'm really on the fence about that one point, and it's an easy change, so if anyone has thoughts on the matter, chime in.  Beyond that, this is done.

stevededalus commented on api/python/tests/test_formats.py:
```diff
@@ -0,0 +1,84 @@
+    meta = {}
+    assert df.equals(FormatsRegistry.deserialize(FormatsRegistry.serialize(df, meta), meta))
+
+
```
extra white space here?

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,415 @@
+        if deserializer:
+            self._deserializer = deserializer
+
+        if not (getattr(self, '_serializer', None) or (self.serialize != Format.serialize)):
```
Can you refactor this? I'd really prefer an abstract base class and a subclass instead of attribute manipulation.

dimaryaz commented on api/python/tests/test_formats.py:
```diff
@@ -0,0 +1,78 @@
+#! /usr/bin/python
```
Still not fixed?

eode replied:
:-(  I removed that..  ..ah.  I thought your comment was in 'formats.py'.

eode replied:
> So anyways... Until we switch to proper URI classes, can you add the unquote calls?

Definitely.

eode replied:
> Can you refactor this?

..hokay.

eode replied:
> Refactor

To be clear, we'll still probably want to do *something* to enable creation of simple formats by users.  That was made that way so that one call could create a format in cases where serialization / deserialization isn't a complex matter at all (JSON, Unicode, yaml, and similar formats that don't have a lot of underlying complexity and variance like CSV and some others do.  ..you have any complaints about a class factory function, then?

eode commented:
Added Pandas<-->CSV format. 
Fixed Parquet settings to be comaptible with s3 select (disabled whole-file compression, enabled per-column compression on all columns)

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,505 @@
+        # add user extensions if given
+        handled_extensions = [] if handled_extensions is None else handled_extensions
+        self.handled_extensions = getattr(self, 'handled_extensions', [])
+        self.handled_extensions.extend(ext for ext in handled_extensions if ext not in self.handled_extensions)
```
This is O(N²)... but really, I'd rather get rid of the `getattr` above.

eode replied:
Sets are overkill, but sure.

eode replied:
done

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,505 @@
+        # add user types if given
+        handled_types = [] if handled_types is None else handled_types
+        self.handled_types = getattr(self, 'handled_types', [])
+        self.handled_types.extend(ext for ext in handled_types if ext not in self.handled_types)
```
Same here

eode replied:
done

eode commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,505 @@
+        Returns:
+            bool
+        """
+        exts = [e.lstrip().lower() for e in self.handled_extensions]
```
Ok, I'll make `handled_extensions` private, too then.

eode replied:
On second thought, I don't know.  We do want users to be able to see easily what extensions a format handles (if they have a format object for whatever reason).  ..and if they see that, they might add to it.  I could make it a property that returns a copy, but that seems like overkill.

eode replied:
I suppose we'll just rely on users not modifying it (or modifying it properly) and leave it public.

eode replied:
done

stevededalus commented on api/python/tests/test_formats.py:
```diff
@@ -71,14 +74,74 @@ def test_formats_serdes():
+    df2 = FormatRegistry.deserialize(bin, meta)
+
+    assert test_data == bin
+    assert df1.equals(df2)
```
new line at end of file

eode commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,1001 @@
+                'nonnumeric': csv.QUOTE_NONNUMERIC
+            }
+            return map.get(value, NOT_SET)
+        print("Unrecognized value for 'quoting' option: {!r}".format(value))
```
Well, it's possible we'll get bad or outdated options in the future -- from someone messing with metadata, or from another client (if we do one in R, or similar).  I've changed it to a warning -- good?

..if there's a problematic or unknown option, this allows it to pass through and the serialization/deserialization to continue.

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,1007 @@
+        handlers.insert(0, handler)
+
+    @classmethod
+    def search(cls, obj_type=None, meta=None, ext=None, single=True, handlers=None):
```
Please remove `single` from all the functions - it's weird to have a return type that depends on arguments.

If the list return type gets annoying, we can add helper functions that return a single item.

eode replied:
Done.

dimaryaz commented on api/python/t4/formats.py:
```diff
@@ -0,0 +1,1007 @@
+        return result_kwargs
+
+
+    def serialize(self, obj, meta=None, ext=None, raw_args=None, **format_opts):
```
Remove `raw_args`; we can add it back if it comes up. It's safer to start with a minimal interface, and add things later.

eode replied:
Done.

eode commented on api/python/t4/api.py:
```diff
@@ -72,17 +72,11 @@ def get(src):
-        raise QuiltException("No serialization metadata")
+    clean_src = fix_url(src)
+    data, meta = get_bytes(clean_src)
+    ext = pathlib.PurePosixPath(urlparse(unquote(clean_src)).path).suffix
```
Good?

dimaryaz commented on api/python/t4/packages.py:
```diff
@@ -189,20 +195,21 @@ def deserialize(self):
+        if func is not None:
+            return func(data)
+
+        pkey_ext = pathlib.PurePosixPath(urlparse(unquote(physical_key)).path).suffix
```
Fix this unquote, too: `unquote(urlparse(physical_key).path)`

eode replied:
Ugh.  I've been awful about those.

eode replied:
Fixed, checked full changeset for any others, found one, fixed that as well.
